fix(compiler): fix a few bugs with the newly introduced `**` operator

mmalerba opened the pull request:
Some fixes for exponentiation which was recently introduced in https://github.com/angular/angular/pull/59894

- Makes exponentiation right-associative. For example, `a ** b ** c` should be equivalent to `a ** (b ** c)`, not `(a ** b) ** c`
- Ensures that required parentheses are preserved in the compiled output. Parentheses are required when a unary operator is used in the base of the exponentiation, for example `(-1) ** 3`

mmalerba committed:
fix(compiler): exponentiation should be right-to-left associative

For example, `a ** b ** c` should be equivalent to `a ** (b ** c)`,
not `(a ** b) ** c`

mmalerba committed:
fix(compiler-cli): preserve required parens in exponentiation expressions

Parentheses are required around a unary operator used in the base of an
exponentiation expression. For example: `(-1) ** 3`

mmalerba commented on packages/compiler-cli/src/ngtsc/translator/src/translator.ts:
```diff
@@ -395,12 +395,30 @@ export class ExpressionTranslatorVisitor<TFile, TStatement, TExpression>
     );
   }
 
+  private visitBinaryOperand(
```
Q: Is this the right place to do this, or should it be part of the template pipeline?

crisbeto replied:
I think it's fine here, but I'm wondering if we need it in other places that translate our AST to something else. E.g. we have a similar translator for converting into the Babel AST in the linker. Also potentially the one that converts to output AST and then JS in JIT mode.

mmalerba replied:
Yeah I was worried about that, it feels like it should be captured in the AST node somehow. There is actually a `parens` property on unary and binary operators, but it looks like its pretty much always `true` and inconsistently respected by the various translators you mentioned.

I guess there are a couple options:
1. Duplicate this logic into all the places we translate ASTs
2. Change the `parens` property to not always just be true, and apply it consistently in all translators
4. Add a `ParenthesizedExpr` to the output AST

Personally I like 3 because that allows us to put parens around any type of expression without needing to add a `parens` property to everything. How breaking is that type of change though? Are there tools that use the AST that will blow up if they aren't updated to know about `ParenthesizedExpr`? Can we make that change one-shot in a single major or does it need to happen over several versions?

crisbeto replied:
IMO it makes sense to have a parenthesized expression, given that's how TypeScript deals with it as well. I'd be concerned that there would be some subtle behavior differences, but it's hard to tell without actually implementing it and running a TGP.

mmalerba replied:
Ok, I'll try it out and if its too difficult to land I'll fallback to `#1`

mmalerba replied:
Opened https://github.com/angular/angular/pull/60127 for it, results look good so far

mmalerba replied:
updated to use `ParenthesizedExpr`

crisbeto commented on packages/compiler-cli/src/ngtsc/translator/src/translator.ts:
```diff
@@ -395,12 +395,30 @@ export class ExpressionTranslatorVisitor<TFile, TStatement, TExpression>
+  ): TExpression {
+    if (
+      isLeftSide &&
+      operator === o.BinaryOperator.Exponentiation &&
```
Should we have a flag on the AST node itself saying if it's right-associative or not? Otherwise we may have to duplicate this logic in other places.

mmalerba committed:
fix(compiler): proper handling of typeof, void in RecursiveAstVisitor

Handle typeof and void expressions the same way as other unary operator
expressions.

mmalerba commented:
This PR was merged into the repository by commit b70ad3c4e63158a72b8aea173b1268ec8ab08e2b.

The changes were merged into the following branches: main
